[FLINK-1297] Added OperatorStatsAccumulator for tracking operator related stats

tammymendt opened the pull request:
The accumulator tracks min and max values, and estimates for count distinct and heavy hitters.

The count distinct algorithms are Linear Counting and HyperLogLog, both from an imported library (clearspring).

The heavy hitters algorithms are Lossy counting (Manku et.al 2002) and Count Min Sketch (Cormode 2005).

The heavy hitters algorithms are implemented in the statistics package in flink-core.

The accumulator currently only uses Linear Counting as default for count distinct and Lossy Counting as default for heavy hitters. 

The accumulator does not only track the globally merged value the way the other accumulators do. It additionally tracks an array of local statistics which have been collected at each subtask of a task. It does this by wrapping an extra class called OperatorStatisticsResult which holds the local and global accumulated results. The idea of this is to be able to track statistics of data processed in subtasks, so that they can be used to reason about partitioning strategies.

aalexandrov commented:
:+1: thanks for the great work! I'll review that (probably over the weekend) and will appreciate if some of the core Flink committers (@sewen, @rmetzger, @fhueske) can also make a pass over the code.

One more caveat from me: this implements only the runtime aspect of the statistics collecting logic.
A second PR which allows to configure the points where statistics should be tracked in a programmatic way as part of the DataBag API shoud follow.

@tammymendt and me were discussing as syntax along the lines of:

```scala
A = // some dataflow assembly code
A.withStatistics( "statsForX", keySelectorFn ) 

env.execute()

// grab the statistics after the execution is done
env.getAccumulator("statsForX")
```

Once this is in place we will play around and implement some ideas on incremental optimization.

fhueske commented:
Nice, statistics collection :-)

I'll have a look at this PR. 
It might take a few days as I am traveling.

rmetzger commented on flink-core/pom.xml:
```diff
@@ -63,6 +63,19 @@ under the License.
 			<version>${guava.version}</version>
 		</dependency>
+
+        <dependency>
```
Seems to be ASL 2.0 (https://github.com/addthis/stream-lib/blob/master/LICENSE.txt) --> good

rmetzger commented on flink-core/src/main/java/org/apache/flink/api/common/accumulators/OperatorStatsAccumulator.java:
```diff
@@ -0,0 +1,98 @@
+		try {
+			localValue.setLocal((OperatorStatistics) in.readObject());
+		} catch (ClassNotFoundException e) {
+			e.printStackTrace();
```
Just doing e.printStackTrace() will make it extremely hard to debug the error. You'll find a stacktrace without any message in the stderr file ... The flink programm will continue running in an erroneous state.
I would recommend re-throwing the exception as a runtime exception.

rmetzger commented on flink-tests/src/test/java/org/apache/flink/test/accumulators/OperatorStatsAccumulatorsTest.java:
```diff
@@ -0,0 +1,118 @@
+			Assert.assertEquals(merged.getHeavyHitters().size(),globalStats.getHeavyHitters().size());
+
+		}
+		catch (Exception e) {
```
I would fail the test in case of an exception

uce replied:
Tests fail anyways if an Exception is thrown. No need for the catch => fail.

rmetzger commented on flink-core/src/test/java/org/apache/flink/statistics/heavyhitters/LossyCountingTest.java:
```diff
@@ -0,0 +1,129 @@
+			actualFreq[x]++;
+		}
+
+		System.out.println("Size of heavy hitters: "+lossyCounting.heavyHitters.size());
```
can you use regular logging instead of system.outs for these messages?
We have different logging configurations for travis, local and IDE but we can not control system.outs with our logging settings.

rmetzger commented on flink-core/src/main/java/org/apache/flink/statistics/OperatorStatistics.java:
```diff
@@ -0,0 +1,154 @@
+		try {
+			this.heavyHitter.merge(other.heavyHitter);
+		} catch (HeavyHitterMergeException e) {
+			e.printStackTrace();
```
rethrow

rmetzger commented on flink-core/src/main/java/org/apache/flink/statistics/OperatorStatistics.java:
```diff
@@ -0,0 +1,154 @@
+			ICardinality mergedCountDistinct = this.countDistinct.merge(new ICardinality[]{this.countDistinct,other.countDistinct});
+			this.countDistinct = mergedCountDistinct;
+		} catch (CardinalityMergeException e) {
+			e.printStackTrace();
```
same here

rmetzger commented:
I made some comments to the PR.
You can update it by pushing into the same branch.

StephanEwen commented on flink-core/pom.xml:
```diff
@@ -63,6 +63,19 @@ under the License.
+            <exclusions>
+                <exclusion>
+                    <groupId>it.unimi.dsi</groupId>
+                    <artifactId>fastutil</artifactId>
```
We have so far refrained from including fastutil, as it is a huge dependency, and one usually needs only a minimal subset of the functionality. Can you elaborate what this is needed for?

rmetzger replied:
This is an exclusion of fastutil ;)

tammymendt replied:
The library we are using from clearspring implements a series of streaming algorithms. We are only using HyperLogLog and CountMinSketch to estimate count distinct and heavy hitters respectively. As far as I could see fastutil is used only in the implementation of a separate streaming algorithm to track histograms (QDigest). Since we are not using this particular algorithm, it should be ok to exclude fastutil from the dependency.

StephanEwen replied:
Ignore my comment. Robert is right, I overlooked that it is an exclusion, not a dependency...

StephanEwen commented:
I think this is nice work.

It would be easier to add and maintain this, if it were not hardwired into the core flink classes, but if we could make this like a library. This would be a good example for the `flink-contrib` project.

The changes for this should be rather small: It would boil down to not adding the specialized methods for the `OperatorStatistics` to the RuntimeEnvironment and the `JobExecutionResult`, but to add a util class that converts an operator statistics accumulator into the OperatorStatistics object.

I general, I am a bit more comfortable with adding stuff as libraries than to the core API. A lot of requests and features are coming in, if we all add them to the core API, it will easily become unmaintainable. Staging something first as a library is pretty uncritical. We can always add it to the API later, if it becomes stable and heavily used.

rmetzger commented:
+1 for Stephans suggestion to make it a library in flink-contrib.

tammymendt commented:
Hey, thanks for the quick feedback. I agree with separating the contribution as a library, however I have a question. The reason I have used a specialized method in the RuntimeEnvironment, is that for the OperatorStatsAccumulator, we do not want to only track the accumulated statistics per job, but also keep an array of the statistics collected at each task. The subtaskIndex which part of the RuntimeEnvironment is passed as a parameter to the constructor of the accumulator so that we can track which subtask is associated to which stats. Any ideas as to how I could use the same accumulator infrastructure but be able to associate an accumulated value with a given subtask?

StephanEwen commented:
Keeping the accumulators from each task separately is something that we want to change anyways in the main system. Then, this would not need special casing.

Until then, you should be able to encode the subtask into the accumulator name, like "taskname_i". Then, the accumulators do not get combined.

fhueske commented on flink-core/src/main/java/org/apache/flink/statistics/heavyhitters/IHeavyHitter.java:
```diff
@@ -0,0 +1,37 @@
+ * Interface for classes that track heavy hitters. It follows the same design as the interfaces from
+ * {@link com.clearspring.analytics.stream}
+ */
+public interface IHeavyHitter {
```
Interfaces do not follow a special naming convention in the Flink code base such as starting with a capital `I`. Please rename to `HeavyHitter` for consistency.

fhueske commented on flink-core/src/main/java/org/apache/flink/statistics/OperatorStatistics.java:
```diff
@@ -0,0 +1,154 @@
+		}
+	}
+
+	public void process(Object tupleObject){
```
It looks like this method expected to be called for each passing element if statistics collection is enabled. This could add significant processing overhead. Would it make sense to add an optional skip interval n that processes only every n-th element (except from increasing the `cardinality` counter.

Also it makes sense to perform as many checks outside of the function as possible. The type of the data elements is known before execution. So it can be checked if a data type allows statistics collection at program composition or optimization time.

tammymendt replied:
The problem with processing only every n-th element is that precision would be affected, and the algorithms are already estimating as it is. I am planning to measure performance overhead but haven't done it so far. 

I have removed unnecessary checks inside the "process" function. The point of them was to allow the user to configure which stats wished to be tracked, by means of the OperatorStatisticsConfig class. This way for example an OperatorStatisticsAccumulator could be defined, but only used to track count distinct. Do you think this makes sense, or is it preferable to create multiple accumulators, one for each type of statistic? I was avoiding this because the common case would be to want all 4 statistics (max, min, heavy hitters and count distinct), and I didn't want the user to have to create 4 separate accumulators for this purpose.

tammymendt commented on flink-contrib/src/test/java/org/apache/flink/contrib/operatorstatistics/OperatorStatsAccumulatorsTest.java:
```diff
@@ -0,0 +1,144 @@
+
+		// Is instantiated later since the runtime context is not yet initialized
+		private Accumulator<Object, Serializable> globalAccumulator;
+		private Accumulator<Object,Serializable>[] localAccumulators;
```
This is the solution I have found for temporarily storing local, as well as global accumulators. But I think there should be a better way to do this, rather than having the user have to specify it and later update both inside of the UDF.

aalexandrov replied:
Does it make sense to use some event notification logic for pluggable components at critical points (e.g. open/close method of UDFs)?

StephanEwen replied:
Can you elaborate what you mean by "event notification logic for pluggable components"? I am a bit confused here ;-)

aalexandrov replied:
I'm sorry, I think the comment was out of the context of the code above (my bad).

I was proposing to expose events from the lifecycle of of an operator (e.g., `Open(Operator)`, `Close(Operator)`). If these are created on points out of the hot path during execution, the performance penalty should not be too high. Add-on components can then react on the events with listeners.

In the statistics collection, we can have a listener that sets up the accumulator when statistics tracking has been configured for an operator. Couple this with a Scala trait in the add-ons package, and we can have a fluent API that sets up statistics tracking:

```scala
// p is a projection function for tracking statistics
// alleviates the need to use RichMapFunction for f
val X = Y.map(f).withStatistics("Y-stats", k)
```

StephanEwen commented:
I am merging this for the next version.
Very nice addition, sorry for the delay.

StephanEwen commented:
The tests seem to be non-deterministic and fail frequently.

Check out this build: https://travis-ci.org/StephanEwen/incubator-flink/jobs/65634990

The tests need to be more stable before we can add this to the codebase.

tammymendt commented:
Ok, I will check this out today.

tammymendt commented:
Hey! 
So I solved the NullPointer and as far as I can see it is not bothering any more. Given the non-deterministic nature of the algorithms, some tests fail sometimes. I am looking into this. I am considering  leaving out tests for the accuracy of the algorithms (in the Accumulator test), given that they are already programmed in other test classes. This should avoid these "algorithm accuracy" related errors in the OperatorStatisticsAccumulatorTest class.

Also, I have been refactoring this code to avoid using the conditionals that check which statistics are being collected in the "process" and "merge" methods. This is based on a remark that @fhueske made way earlier in the PR. Should I try and make a PR for the refactored code, or rather leave this in its current state, and wait until it is merged until I make a PR for the refactored version?

Cheers!

tammymendt commented:
Everything seems to be OK now. Is there still a chance to merge this before the upcoming release?

I have pushed two commits. The first is the original version with the bug fixes. The second is a re-factored version that makes the code more flexible and extensible. The second version has a couple more classes than the first though. I have pushed both in separate commits so you can decide which version you think is preferable. 

Cheers,

Tamara

fhueske commented:
Hi @tammymendt 

thanks for the update! I'm afraid, we are already in the testing phase for the release and don't want to add new features. Since we postponed a few important features to the next release, we will push hard to get the next release out rather soon.

I'll have a look at the PR in detail after the release stress is over.

Fabian

tammymendt commented:
Sounds good! Thanks =)

tammymendt commented:
Hey! So I've been using and testing this code throughout my master thesis. Collecting count distinct makes jobs about 10% slower whereas collecting heavy hitters can make a job be 20 to 50% slower (depending on the algorithm and the distribution of the data). However this overhead is lower than that of using a histogram accumulator (not to mention the histogram might not fit in memory). I think it can be a nice addition to the code, specially since it does not affect any core components. 

The version that I pushed now uses a bunch of conditionals to check which statistic is being collected. I know @fhueske did not really like this. I implemented another version which avoids the conditionals by using a different class for every type of statistic. I preferred to push this version though, since it has been more thoroughly tested.

StephanEwen commented:
It is a nice addition, I agree.

The question for all code that we add is who can maintain it. Code never gets just added and then simply works and can stay. Any code that is added needs to be maintained by someone, be that a committer, or by a commitment from the contributor.

If we can find a maintainer, than I am find with adding this.

aalexandrov commented:
@StephanEwen To be honest I'm kind of puzzled and somewhat annoyed that a PR that

1) adds a feature that has been on the list for at least 2 years,
2) has been tested and evaluated thoroughly, and
3) has been lying around since April and was rebased at least three times

is still open. 

I was also not aware that there is a strict maintainer policy for `flink-contrib` commits. IMHO this is a good place to expose non-critical / exploratory / unstable features that might be of benefit for multiple users to the project. If these turn out the be useful for enough people, I'm sure there will be incentive to maintain them. Otherwise there is always the option to deprecate and cleanup recent additions to the contrib prior to the next release if nobody picked up on them.

lauritzthamsen commented:
I would also like to see this merged. I'm planning to use it and afaik some other users too.

I think having it in flink-contrib would be nice because a. it's more visible than in a user's fork/repository, b. everybody can contribute back improvements here (via PRs and some more eyes of the Flink community) and c. it'll at least receive the automatic changes from tool-based modifications in Flink's core.

tammymendt commented:
I will maintain the code. Though I am not committing other code to Flink, I am responsible for what I coded and will not bail as soon as it is merged. I think this feature might be useful for other people, but if you prefer not to merge it I can close the PR and just leave it on my branch.

fhueske commented:
I think there is little risk in merging it to `flink-contrib`. We made that place exactly for this contributions like this one. If we find that the code is popular and used by many people, we can also move it.

In general, I am +1 for adding this feature.

StephanEwen commented:
Yes, if there is commitment to to maintain the code and not just "dump" it, it is good to merge.

rmetzger commented on flink-contrib/flink-operator-stats/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        </dependency>
+
+        <dependency>
+            <groupId>com.clearspring.analytics</groupId>
```
ASL 2.0 license, that's good.

rmetzger commented:
+1 for merging.

fhueske commented:
Hi @tammymendt, it's been a while again. Sorry!
I'll try to rebase and merge this PR tomorrow.

fhueske commented:
After rebasing the PR to the current master, the `OperatorStatsAccumulatorTest` is failing. In the original PR (based on a master from begin of August) the test is passing. @mxm, you are more familiar with Flink's accumulators. Can you have a look? Thanks!

tammymendt commented:
I can have a look at it this afternoon. I suspect it might be because the
function getAllAccumulators has been deprecated.
On Sep 9, 2015 11:11 AM, "Fabian Hueske" <notifications@github.com> wrote:

> After rebasing the PR to the current master, the
> OperatorStatsAccumulatorTest is failing. In the original PR (based on a
> master from begin of August) the test is passing. @mxm
> <https://github.com/mxm>, you are more familiar with Flink's
> accumulators. Can you have a look? Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/605#issuecomment-138847826>.
>

mxm commented:
Hi @tammymendt. Thanks for the pull request! The accumulators work a little bit different now because they are now accumulated on a per-task basis and reported to the job manager in regular intervals.

The `clone()` method in `OperatorStatistics` doesn't create a deep copy of the object, i.e. some references are reused. That causes problems when merging the accumulators because runtime accumulators are modified while merging accumulators for sending them to the job manager.

I could make the test pass by a nasty deep copy using Java serialization. However, I didn't managed to make a proper copy using the provided interfaces. I think you can probably do that faster because you know the code very well.

mxm commented:
I tried again, this works:

```java
	@Override
	public OperatorStatistics clone(){
		OperatorStatistics clone = new OperatorStatistics(config);
		clone.min = min;
		clone.max = max;
		clone.cardinality = cardinality;

		try {
			ICardinality copy;
			if (countDistinct instanceof LinearCounting) {
				copy = new LinearCounting(config.getCountDbitmap());
			} else if (countDistinct instanceof HyperLogLog) {
				copy = new HyperLogLog(config.getCountDlog2m());
			} else {
				throw new IllegalStateException("Unsupported counter.");
			}
			clone.countDistinct = copy.merge(countDistinct);
		} catch (CardinalityMergeException e) {
			throw new RuntimeException("Faild to clone OperatorStatistics!");
		}

		try {
			HeavyHitter copy;
			if (heavyHitter instanceof LossyCounting) {
				copy = new LossyCounting(config.getHeavyHitterFraction(), config.getHeavyHitterError());
			} else if (heavyHitter instanceof CountMinHeavyHitter) {
				copy = new CountMinHeavyHitter(config.getHeavyHitterFraction(),
						config.getHeavyHitterError(),
						config.getHeavyHitterConfidence(),
						config.getHeavyHitterSeed());
			} else {
				throw new IllegalStateException("Unsupported counter.");
			}
			copy.merge(heavyHitter);
			clone.heavyHitter = copy;
		} catch (HeavyHitterMergeException e) {
			throw new RuntimeException("Failed to clone OperatorStatistics!");
		}

		return clone;
	}
```

Do you think we could merge your pull request with this change?

tammymendt commented:
The test I wrote checks if the final accumulator result is the one that is expected. If I understand correctly, now accumulator values are sent periodically before a task finishes. Do I need to write an additional test to make sure that accumulator results before task end are also as expected? If not I would say it is ok to merge. Have you written extra tests for other accumulators like IntCounter?

mxm commented:
Yes, there are tests in place of the new accumulators. The included Accumulators all implement a deep copy in the clone() method. That's why they didn't have to be ported. You don't need to include an extra test for the runtime accumulators.

mxm commented:
Could you update the pull request with the new `clone()` method? I will then merge the pull request if it passes Travis.

tammymendt commented:
Yes, I'm on it

On Wed, Sep 9, 2015 at 2:34 PM, Max <notifications@github.com> wrote:

> Could you update the pull request with the new clone() method? I will
> then merge the pull request if it passes Travis.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/605#issuecomment-138895569>.
>



-- 
Tamara Mendt

tammymendt commented:
I rebased with master and included the new clone() method. I had to include extra conditionals that check whether count distinct or heavy hitters are being tracked because otherwise the clone method would throw an exception in some cases in which it shouldn't. Also, I added some tests to the OperatorStatsAccumulatorTest class because I realized that I had not written tests for different configurations of the OperatorStatsConfig class.

mxm commented:
Looks good. Merging this now.

fhueske commented:
Thanks @tammymendt for the PR and the last fixes. Sorry that it took us so long to merge it :-/
